In [3]:
import numpy as np

# Generate a random 5x4 matrix with linearly independent columns
random_matrix = np.random.rand(5, 4)

# Perform QR decomposition
Q, R = np.linalg.qr(random_matrix)

# Display the original matrix, Q, and R
print("Original Random Matrix:")
print(random_matrix)
print("\nMatrix Q (Orthogonal):")
print(Q)
print("\nMatrix R (Upper Triangular):")
print(R)

# Observations on the diagonal elements of R
diagonal_elements_R = np.diagonal(R)
print("\nDiagonal Elements of R:", diagonal_elements_R)



Original Random Matrix:
[[0.61478852 0.33742584 0.24682943 0.77071764]
 [0.10262585 0.52205792 0.35714524 0.55204712]
 [0.71200689 0.52813722 0.83926185 0.30423309]
 [0.37170876 0.77398274 0.85555838 0.99352545]
 [0.45366106 0.34659025 0.73136953 0.68180559]]

Matrix Q (Orthogonal):
[[-0.55222409  0.30042737  0.7032628  -0.32923019]
 [-0.09218205 -0.65200575  0.31097338  0.13688198]
 [-0.63954896  0.14085575 -0.22415899  0.72170455]
 [-0.33388153 -0.67766845 -0.09897057 -0.18949277]
 [-0.40749389  0.07454547 -0.59048671 -0.5622329 ]]

Matrix R (Upper Triangular):
[[-1.11329536 -0.97188061 -1.28966034 -1.28062106]
 [ 0.         -0.6632886  -0.56575621 -0.70799558]
 [ 0.          0.         -0.42001856  0.1445655 ]
 [ 0.          0.          0.         -0.53021123]]

Diagonal Elements of R: [-1.11329536 -0.6632886  -0.42001856 -0.53021123]


Anshuman 

Original Random Matrix:
[[0.61478852 0.33742584 0.24682943 0.77071764]
 [0.10262585 0.52205792 0.35714524 0.55204712]
 [0.71200689 0.52813722 0.83926185 0.30423309]
 [0.37170876 0.77398274 0.85555838 0.99352545]
 [0.45366106 0.34659025 0.73136953 0.68180559]]

Matrix Q (Orthogonal):
[[-0.55222409  0.30042737  0.7032628  -0.32923019]
 [-0.09218205 -0.65200575  0.31097338  0.13688198]
 [-0.63954896  0.14085575 -0.22415899  0.72170455]
 [-0.33388153 -0.67766845 -0.09897057 -0.18949277]
 [-0.40749389  0.07454547 -0.59048671 -0.5622329 ]]

Matrix R (Upper Triangular):
[[-1.11329536 -0.97188061 -1.28966034 -1.28062106]
 [ 0.         -0.6632886  -0.56575621 -0.70799558]
 [ 0.          0.         -0.42001856  0.1445655 ]
 [ 0.          0.          0.         -0.53021123]]

Diagonal Elements of R: [-1.11329536 -0.6632886  -0.42001856 -0.53021123]

Amit Bhai 

Original Random Matrix:
[[0.62285974 0.3956732  0.78942982 0.71206619]
 [0.21538468 0.19833982 0.70422911 0.51401903]
 [0.61374532 0.30540896 0.27383987 0.24551591]
 [0.21101708 0.86485664 0.29597694 0.36761226]
 [0.59161111 0.41841235 0.4006921  0.29453886]]

Matrix Q (Orthogonal):
[[-0.56727787  0.09989822  0.37236933 -0.65701356]
 [-0.19616449 -0.049925    0.79665877  0.44262973]
 [-0.55897679  0.21443068 -0.40901071 -0.04386259]
 [-0.19218663 -0.96965046 -0.10142348 -0.07751236]
 [-0.53881775  0.0364048  -0.22158385  0.6037218 ]]

Matrix R (Upper Triangular):
[[-1.09797998 -0.82574228 -1.01182377 -0.87136228]
 [ 0.         -0.72826244 -0.1699834  -0.24761489]
 [ 0.          0.          0.62418041  0.47168117]
 [ 0.          0.          0.         -0.10176097]]

Diagonal Elements of R: [-1.09797998 -0.72826244  0.62418041 -0.10176097]

Vishal Bhai 

Original Random Matrix:
[[0.49514283 0.01214133 0.99681964 0.71543333]
 [0.94951097 0.20958943 0.62634127 0.57828438]
 [0.50838857 0.57437887 0.76972014 0.73338807]
 [0.71465945 0.75665507 0.62276052 0.65129866]
 [0.48978091 0.4903438  0.36527196 0.65768895]]

Matrix Q (Orthogonal):
[[-0.33722848 -0.42938467  0.75754697  0.10439731]
 [-0.64668643 -0.53909875 -0.462769   -0.10150511]
 [-0.34624981  0.41662921  0.42000288 -0.08020734]
 [-0.48673537  0.50819432 -0.09863104 -0.47328264]
 [-0.33357663  0.30522118 -0.16073875  0.8650828 ]]

Matrix R (Upper Triangular):
[[-1.46827106 -0.87036963 -1.43268358 -1.40556839]
 [ 0.          0.65529144 -0.01701887  0.21832928]
 [ 0.          0.          0.66843417  0.41243301]
 [ 0.          0.          0.          0.21787444]]

Diagonal Elements of R: [-1.46827106  0.65529144  0.66843417  0.21787444]

<h1>Observations on the output:</h1>

<h2> Matrix Q (Orthogonal): </h2>
<p>
   The columns of Q form an orthogonal basis.
   The dot product of any two columns is approximately zero, indicating orthogonality.
</p>


<h2> Matrix R (Upper Triangular): </h2>
<p>
    R is an upper triangular matrix with all entries below the main diagonal being zero.
    The diagonal elements of R are −0.98029322,−0.6544642,0.4355922,−0.3848973−0.98029322,−0.6544642,0.4355922,−0.3848973.
</p>


<h2> Diagonal Elements of R: </h2>
<p>
    The diagonal elements of R represent the scale or magnitude of the corresponding columns of the original matrix.
    All diagonal elements are non-zero, indicating that the original matrix has linearly independent columns.
</p>

<p> 
In summary, the output confirms that the original random 5x4 matrix has been successfully decomposed into an orthogonal matrix
Q and an upper triangular matrix R. The non-zero diagonal elements of 
R indicate the scale of the original columns, and the negative sign in the diagonal elements is common due to the Gram-Schmidt process and doesn't affect the orthogonality of Q.
</p>

Original Random Matrix:
[[0.70027017 0.38686066 0.54336805 0.8118312 ]
 [0.10821299 0.54472045 0.31354576 0.34784126]
 [0.07126129 0.34407601 0.16734807 0.61734904]
 [0.287223   0.55528458 0.92981509 0.88462884]
 [0.60935311 0.6011572  0.65912585 0.73812479]]

Matrix Q (Orthogonal):
[[-0.71434766  0.38891876 -0.10957834 -0.23597703]
 [-0.11038838 -0.68087087 -0.39475321  0.40066192]
 [-0.07269385 -0.42600665 -0.32247602 -0.81341232]
 [-0.29299702 -0.44648772  0.8379588  -0.09016985]
 [-0.6216029  -0.06575795 -0.16123511  0.33766057]]

Matrix R (Upper Triangular):
[[-0.98029322 -0.89787382 -1.11707827 -1.38121896]
 [ 0.         -0.6544642  -0.53194332 -0.62760689]
 [ 0.          0.          0.4355922   0.19692006]
 [ 0.          0.          0.         -0.3848973 ]]

Diagonal Elements of R: [-0.98029322 -0.6544642   0.4355922  -0.3848973 ]
